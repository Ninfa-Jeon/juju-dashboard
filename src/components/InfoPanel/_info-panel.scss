@import "../../scss/settings";
@import "vanilla-framework/scss/vanilla";

.info-panel {
  margin-right: 1rem;
  padding: 1rem;

  &__pictogram {
    background: $color-light;
    border: solid 1px $color-mid-x-light;
    margin-bottom: 1rem;
  }

  &__grid {
    display: grid;
    grid-template-columns: 1fr;

    @media (min-width: $breakpoint-m-large) {
      grid-template-columns: 1fr 1fr;
    }
  }

  .p-muted-heading {
    font-size: 0.75rem;
    margin-bottom: 0.25rem;
  }
}
