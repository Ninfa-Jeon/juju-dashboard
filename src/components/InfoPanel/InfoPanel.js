import { useState } from "react";
import { useParams } from "react-router-dom";

import Topology from "components/Topology/Topology";
import Modal from "@canonical/react-components/dist/components/Modal";

import { extractCloudName, getViewportWidth } from "app/utils";
import useAnalytics from "hooks/useAnalytics";
import useModelStatus from "hooks/useModelStatus";
import useEventListener from "hooks/useEventListener";

import fullScreenIcon from "static/images/icons/fullscreen-icon.svg";

import "./_info-panel.scss";

const expandedTopologyDimensions = () => {
  const de = document.documentElement;
  const vw = Math.max(de.clientWidth, window.innerWidth || 0);
  const vh = Math.max(de.clientHeight, window.innerHeight || 0);
  const delta = 300;
  return {
    width: vw - delta,
    height: vh - delta,
  };
};

const infoPanelDimensions = () => {
  const vw = getViewportWidth();
  // If changes are made to the 1580px breakpoint below then be sure to update
  // the same breakpoint in _info-panel.scss.
  const size = vw >= 1580 ? 300 : 180;
  return size;
};

const InfoPanel = () => {
  const { 0: modelName } = useParams();
  const [showExpandedTopology, setShowExpandedTopology] = useState(false);
  const modelStatusData = useModelStatus();
  const applicationsCount = Object.entries(modelStatusData.applications || {})
    .length;

  const cloudProvider = modelStatusData
    ? extractCloudName(modelStatusData.model["cloud-tag"])
    : "";

  const { width, height } = expandedTopologyDimensions();
  const topologySize = infoPanelDimensions();

  const sendAnalytics = useAnalytics();

  // Close topology, if open, on Escape key press
  const closeOnEscape = function (e) {
    if (e.code === "Escape" && showExpandedTopology) {
      setShowExpandedTopology(false);
    }
  };
  useEventListener("keydown", closeOnEscape);

  const handleExpandTopology = () => {
    setShowExpandedTopology(!showExpandedTopology);
    sendAnalytics({
      category: "User",
      action: "Opened expanded topology",
    });
  };

  return (
    <div className="info-panel">
      {showExpandedTopology ? (
        <Modal
          close={() => setShowExpandedTopology(false)}
          title={modelName.split("/")[1] || modelName}
          data-test="topology-modal"
        >
          <Topology width={width} height={height} modelData={modelStatusData} />
        </Modal>
      ) : (
        <>
          {applicationsCount > 0 && (
            <div className="info-panel__pictogram">
              <Topology
                width={topologySize}
                height={topologySize}
                modelData={modelStatusData}
                data-test="topology"
              />
              {modelName !== undefined && (
                <i
                  // @TODO the .p-icon--expand class can be removed when this issue lands
                  // https://github.com/canonical-web-and-design/jaas-dashboard/issues/453
                  className="p-icon--expand p-icon--fullscreen"
                  style={{ backgroundImage: `url(${fullScreenIcon})` }}
                  onClick={handleExpandTopology}
                  onKeyPress={handleExpandTopology}
                  role="button"
                  tabIndex="0"
                >
                  Expand topology
                </i>
              )}
            </div>
          )}
        </>
      )}
      <div className="info-panel__grid">
        <div className="info-panel__grid-item">
          <h4 className="p-muted-heading">Controller</h4>
          <p data-name="controller">
            {modelStatusData ? modelStatusData.model.type : ""}
          </p>
        </div>
        <div className="info-panel__grid-item">
          <h4 className="p-muted-heading">Cloud/Region</h4>
          <p data-name="cloud-region">
            {cloudProvider}
            {modelStatusData ? "/" : ""}
            {modelStatusData ? modelStatusData.model.region : ""}
          </p>
        </div>
        <div className="info-panel__grid-item">
          <h4 className="p-muted-heading">Version</h4>
          <p data-name="version">
            {modelStatusData ? modelStatusData.model.version : ""}
          </p>
        </div>
        <div className="info-panel__grid-item">
          <h4 className="p-muted-heading">SLA</h4>
          <p data-name="sla">
            {modelStatusData ? modelStatusData.model.sla : ""}
          </p>
        </div>
      </div>
    </div>
  );
};

export default InfoPanel;
