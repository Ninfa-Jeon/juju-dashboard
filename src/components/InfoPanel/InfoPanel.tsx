import { useState } from "react";
import { useParams } from "react-router-dom";

import Topology from "components/Topology/Topology";
import Modal from "@canonical/react-components/dist/components/Modal";

import { extractCloudName, getViewportWidth } from "app/utils/utils";
import useAnalytics from "hooks/useAnalytics";
import useModelStatus from "hooks/useModelStatus";
import useEventListener from "hooks/useEventListener";

import { validateModelNameFromURL } from "app/selectors";

import type { TSFixMe } from "types";

import "./_info-panel.scss";

const expandedTopologyDimensions = () => {
  const de = document.documentElement;
  const vw = Math.max(de.clientWidth, window.innerWidth || 0);
  const vh = Math.max(de.clientHeight, window.innerHeight || 0);
  const delta = 300;
  return {
    width: vw - delta,
    height: vh - delta,
  };
};

const infoPanelDimensions = () => {
  const vw = getViewportWidth();
  // If changes are made to the 1580px breakpoint below then be sure to update
  // the same breakpoint in _info-panel.scss.
  const size = vw >= 1580 ? 300 : 180;
  return size;
};

const InfoPanel = () => {
  let { userName, modelName } = useParams<{
    userName: string;
    modelName: string;
  }>();
  modelName = validateModelNameFromURL(userName, modelName);

  const [showExpandedTopology, setShowExpandedTopology] = useState(false);
  const modelStatusData: TSFixMe = useModelStatus();
  const applicationsCount = Object.entries(modelStatusData.applications || {})
    .length;

  const cloudProvider = modelStatusData
    ? extractCloudName(modelStatusData.model["cloud-tag"])
    : "";

  const { width, height } = expandedTopologyDimensions();
  const topologySize = infoPanelDimensions();

  const sendAnalytics = useAnalytics();

  // Close topology, if open, on Escape key press
  const closeOnEscape = function (e: KeyboardEvent) {
    if (e.code === "Escape" && showExpandedTopology) {
      setShowExpandedTopology(false);
    }
  };
  useEventListener("keydown", closeOnEscape);

  const handleExpandTopology = () => {
    setShowExpandedTopology(!showExpandedTopology);
    sendAnalytics({
      path: undefined,
      category: "User",
      action: "Opened expanded topology",
    });
  };

  return (
    <div className="info-panel">
      {showExpandedTopology ? (
        <Modal
          close={() => setShowExpandedTopology(false)}
          title={modelName?.split("/")[1] || modelName}
          data-test="topology-modal"
        >
          <Topology width={width} height={height} modelData={modelStatusData} />
        </Modal>
      ) : (
        <>
          {applicationsCount > 0 && (
            <div className="info-panel__pictogram">
              <Topology
                width={topologySize}
                height={topologySize}
                modelData={modelStatusData}
                data-test="topology"
              />
              {modelName !== undefined && (
                <i
                  className="p-icon--fullscreen"
                  onClick={handleExpandTopology}
                  onKeyPress={handleExpandTopology}
                  role="button"
                  tabIndex={0}
                >
                  Expand topology
                </i>
              )}
            </div>
          )}
        </>
      )}
      <div className="info-panel__grid">
        <div className="info-panel__grid-item">
          <h4 className="p-muted-heading">Controller</h4>
          <p data-name="controller">
            {modelStatusData ? modelStatusData.model.type : ""}
          </p>
        </div>
        <div className="info-panel__grid-item">
          <h4 className="p-muted-heading">Cloud/Region</h4>
          <p data-name="cloud-region">
            {cloudProvider}
            {modelStatusData ? "/" : ""}
            {modelStatusData ? modelStatusData.model.region : ""}
          </p>
        </div>
        <div className="info-panel__grid-item">
          <h4 className="p-muted-heading">Version</h4>
          <p data-name="version">
            {modelStatusData ? modelStatusData.model.version : ""}
          </p>
        </div>
        <div className="info-panel__grid-item">
          <h4 className="p-muted-heading">SLA</h4>
          <p data-name="sla">
            {modelStatusData ? modelStatusData.model.sla : ""}
          </p>
        </div>
      </div>
    </div>
  );
};

export default InfoPanel;
