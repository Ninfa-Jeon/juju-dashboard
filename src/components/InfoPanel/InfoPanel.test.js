import React from "react";
import { shallow } from "enzyme";

import InfoPanel from "./InfoPanel";

describe("Info Panel", () => {
  const modelStatus = {
    model: {
      name: "fake-model",
      region: "eu-ireland",
      type: "iaas",
      cloudTag: "cloud-google",
      sla: "Unsupported"
    }
  };

  it("renders without crashing and matches snapshot", () => {
    const wrapper = shallow(<InfoPanel />);
    expect(wrapper).toMatchSnapshot();
  });

  it("displays correct model status info", () => {
    const wrapper = shallow(<InfoPanel modelStatus={modelStatus} />);
    expect(wrapper.find('[data-name="name"]').text()).toStrictEqual(
      "fake-model"
    );
  });
});
