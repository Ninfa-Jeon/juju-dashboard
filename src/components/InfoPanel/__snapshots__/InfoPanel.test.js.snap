// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Info Panel renders without crashing and matches snapshot 1`] = `
<div
  className="info-panel"
>
  <div
    className="info-panel__grid"
  >
    <div
      className="info-panel__grid-item"
      key="Model"
    >
      <h4
        className="p-muted-heading"
      >
        Model
      </h4>
      <p>
        cdk-default
      </p>
    </div>
    <div
      className="info-panel__grid-item"
      key="Region"
    >
      <h4
        className="p-muted-heading"
      >
        Region
      </h4>
      <p>
        us-east-1
      </p>
    </div>
    <div
      className="info-panel__grid-item"
      key="Controller"
    >
      <h4
        className="p-muted-heading"
      >
        Controller
      </h4>
      <p>
        aws-controller
      </p>
    </div>
    <div
      className="info-panel__grid-item"
      key="Cloud"
    >
      <h4
        className="p-muted-heading"
      >
        Cloud
      </h4>
      <p>
        aws
      </p>
    </div>
    <div
      className="info-panel__grid-item"
      key="SLA"
    >
      <h4
        className="p-muted-heading"
      >
        SLA
      </h4>
      <p>
        Unsupported
      </p>
    </div>
  </div>
</div>
`;
